### ISSUE TITLE
Lint `unreachable_pub` got easily confused by multiple items in a single `pub use ...` statement

### ISSUE THREAD

**Issue by imp** [rust-lang/rust#64762]
This code ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6f6d3006c4df35712474dc8fa6ab8442)) trips `unreachable_pub` lint (while being completely legit).

```rust
#![deny(unreachable_pub)]

pub use self::m1::{Item1, Item2};

mod m1 {
    pub use self::m2::{Item1, Item2};
    
    mod m2 {
        pub struct Item1;
        pub struct Item2;
    }
}
```

However, if we break the top-level `pub use ...` statement into two separate statements ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d48e646d8acdb673edf2d2f55e91e982)) - `unreachable_pub` lint convinces itself that the code is, indeed, legit.

```rust
#![deny(unreachable_pub)]

pub use self::m1::Item1;
pub use self::m1::Item2;

mod m1 {

    pub use self::m2::{Item1, Item2};
    
    mod m2 {
        pub struct Item1;
        pub struct Item2;
    }
}
```

**Comment by jhpratt** [comment]
Running into this myself. Right now I'm just `#[allow]`ing it with a link to this issue.

If someone could lead me in the right direction, I'll give fixing this a shot.

**Comment by HeroicKatora** [comment]
It also gets confused if the export happens via a glob, also on a single item.

```
#![deny(unreachable_pub)]

mod m1 {
    pub mod m2 {
        pub use super::Item1;
    }
    
    pub struct Item1;
}

pub use m1::m2::*;
```

**Comment by jyn514** [comment]
Looks like this is because of https://github.com/rust-lang/rust/pull/57922#discussion_r251234202

**Comment by lambinoo** [comment]
Since I claimed it 8 days ago, just a follow up.
I've been able to break the compiler a few times but I'm getting somewhere I think...
Not to a proper solution tho, just more mess, but that means more knowledge!

tl;dr: I haven't given up!

**Comment by petrochenkov** [comment]
How to fix this: https://github.com/rust-lang/rust/issues/82064.

**Comment by lambinoo** [comment]
Closed by #87487
